lib/vtls/schannel.c

```diff
@@ -1574,7 +1594,332 @@ CURLcode Curl_schannel_random(unsigned char *entropy, size_t length)
+                                   const char *ca_file,
+                                   struct connectdata *conn)
+{
+  CURLcode result;
+  struct Curl_easy *data = conn->data;
+  HANDLE ca_file_handle = INVALID_HANDLE_VALUE;
+  LARGE_INTEGER file_size;
+  char *ca_file_buffer = NULL;
+  char *current_ca_file_ptr = NULL;
+  LPCTSTR ca_file_tstr = NULL;
+  size_t ca_file_bufsize = 0;
+  DWORD total_bytes_read = 0;
+  bool more_certs = 0;
+  bool found_certs = 0;
+  size_t END_CERT_LEN;
+
+  do {
+    ca_file_tstr = Curl_convert_UTF8_to_tchar(ca_file);
+    if(!ca_file_tstr) {
+      failf(data,
+            "schannel: invalid path name for CA file '%s': %s",
+            ca_file,
+            Curl_strerror(conn, GetLastError()));
+      result = CURLE_SSL_CACERT_BADFILE;
+      break;
+    }
+
+    /*
+     * Read the CA file completely into memory before parsing it. This
+     * optimizes for the common case where the CA file will be relatively
+     * small ( < 1 MB ).
+     */
+    ca_file_handle = CreateFile(ca_file_tstr,
+                                GENERIC_READ,
+                                0,
+                                NULL,
+                                OPEN_EXISTING,
+                                FILE_ATTRIBUTE_NORMAL,
+                                NULL);
+    if(ca_file_handle == INVALID_HANDLE_VALUE) {
+      failf(data,
+            "schannel: failed to open CA file '%s': %s",
+            ca_file,
+            Curl_strerror(conn, GetLastError()));
+      result = CURLE_SSL_CACERT_BADFILE;
+      break;
+    }
+
+    if(!GetFileSizeEx(ca_file_handle, &file_size)) {
+      failf(data,
+            "schannel: failed to determine size of CA file '%s': %s",
+            ca_file,
+            Curl_strerror(conn, GetLastError()));
+      result = CURLE_SSL_CACERT_BADFILE;
+      break;
+    }
+
+    if(file_size.QuadPart > MAX_CAFILE_SIZE) {
+      failf(data,
+            "schannel: CA file exceeds max size of %u bytes",
+            MAX_CAFILE_SIZE);
+      result = CURLE_OUT_OF_MEMORY;
+      break;
+    }
+
+    ca_file_bufsize = (size_t)file_size.QuadPart;
+    ca_file_buffer = (char *)malloc(ca_file_bufsize+1);
+    if(!ca_file_buffer) {
+      result = CURLE_OUT_OF_MEMORY;
+      break;
+    }
+
+    result = CURLE_OK;
+    while(total_bytes_read < ca_file_bufsize) {
+      DWORD bytes_to_read = (DWORD)(ca_file_bufsize - total_bytes_read);
+      DWORD bytes_read = 0;
+
+      if(!ReadFile(ca_file_handle, ca_file_buffer + total_bytes_read,
+                   bytes_to_read, &bytes_read, NULL)) {
+
+        failf(data,
+              "schannel: failed to read from CA file '%s': %s",
+              ca_file,
+              Curl_strerror(conn, GetLastError()));
+        result = CURLE_SSL_CACERT_BADFILE;
+        break;
+      }
+      if(bytes_read == 0) {
+        /* Premature EOF -- adjust the bufsize to the new value */
+        ca_file_bufsize = total_bytes_read;
+      }
+      else{
```

Comment:
Please add a space between the else and { to stay in line with the rest of schannel.c